File: api/media-service/src/handlers/user_media.py
"""
User media management handlers
"""

import logging

from models import Image
from utils.auth import get_user_id_from_request
from utils.responses import respond_error, respond_success

logger = logging.getLogger(__name__)


def list_user_media():
    """List all media files for the authenticated user"""
    try:
        logger.info("User media list request received")

        # Get user ID from request
        user_id = get_user_id_from_request()
        if not user_id:
            return respond_error("Authentication required", 401)

        # Get all images for the user
        images = Image.get_by_user_id(user_id)

        # Convert to dict format
        media_list = []
        for image in images:
            media_list.append(
                {
                    "id": image.id,
                    "filename": image.filename,
                    "original_name": image.original_name,
                    "file_size": image.file_size,
                    "mime_type": image.mime_type,
                    "width": image.width,
                    "height": image.height,
                    "is_profile": image.is_profile,
                    "created_at": (
                        image.created_at.isoformat() if image.created_at else None
                    ),
                    "description": image.description,
                    "url": f"/api/v1/media/get/{image.filename}",
                }
            )

        logger.info(f"Returned {len(media_list)} media files for user {user_id}")
        return respond_success(
            {"media": media_list, "count": len(media_list)},
            f"Found {len(media_list)} media files",
        )

    except Exception as e:
        logger.error(f"List user media error: {str(e)}")
        return respond_error("Internal server error during media listing", 500)


def get_user_media(target_user_id):
    """Get media files for a specific user (public endpoint)"""
    try:
        logger.info(f"Public media list request for user {target_user_id}")

        try:
            target_user_id = int(target_user_id)
        except (ValueError, TypeError):
            return respond_error("Invalid user ID", 400)

        # Get all active, public images for the user
        # Note: In a real app, you might want to filter based on privacy settings
        images = Image.get_by_user_id(target_user_id)

        # Convert to dict format (limited info for public view)
        media_list = []
        for image in images:
            media_list.append(
                {
                    "id": image.id,
                    "filename": image.filename,
                    "width": image.width,
                    "height": image.height,
                    "is_profile": image.is_profile,
                    "created_at": (
                        image.created_at.isoformat() if image.created_at else None
                    ),
                    "url": f"/api/v1/media/get/{image.filename}",
                }
            )

        logger.info(
            f"Returned {len(media_list)} public media files for user {target_user_id}"
        )
        return respond_success(
            {"media": media_list, "user_id": target_user_id, "count": len(media_list)},
            f"Found {len(media_list)} media files",
        )

    except Exception as e:
        logger.error(f"Get user media error: {str(e)}")
        return respond_error("Internal server error during media retrieval", 500)


def set_profile_image():
    """Set an image as the user's profile image"""
    try:
        logger.info("Set profile image request received")

        # Get user ID from request
        user_id = get_user_id_from_request()
        if not user_id:
            return respond_error("Authentication required", 401)

        # Get image ID from request
        from flask import request

        data = request.get_json()
        if not data:
            return respond_error("JSON data required", 400)

        image_id = data.get("image_id")
        if not image_id:
            return respond_error("Image ID is required", 400)

        try:
            image_id = int(image_id)
        except (ValueError, TypeError):
            return respond_error("Invalid image ID", 400)

        # Find the image
        image = Image.query.filter_by(id=image_id, is_active=True).first()
        if not image:
            return respond_error("Image not found", 404)

        # Check if user owns this image
        if image.user_id != user_id:
            return respond_error(
                "Access denied: you can only set your own images as profile", 403
            )

        # Set as profile image
        image.set_as_profile()

        logger.info(
            f"Profile image set successfully: image {image_id} for user {user_id}"
        )
        return respond_success(
            {
                "id": image.id,
                "filename": image.filename,
                "is_profile": image.is_profile,
            },
            "Profile image set successfully",
        )

    except Exception as e:
        logger.error(f"Set profile image error: {str(e)}")
        return respond_error("Internal server error during profile image setting", 500)

File: api/media-service/src/handlers/user_media.go
package handlers

import (
	"log"
	"strconv"

	"media-service/src/conf"
	"media-service/src/models"
	"media-service/src/utils"

	"github.com/gin-gonic/gin"
)

// ListUserMediaHandler returns all media for the authenticated user
func ListUserMediaHandler(c *gin.Context) {
	// Get user ID from context
	userID, exists := c.Get("user_id")
	if !exists {
		utils.RespondError(c, "Authentication required", 401)
		return
	}

	var images []models.Image
	if err := conf.DB.Where("user_id = ? AND is_active = ?", userID, true).
		Order("created_at DESC").Find(&images).Error; err != nil {
		log.Printf("Failed to fetch user media: %v", err)
		utils.RespondError(c, "Failed to fetch media", 500)
		return
	}

	// Convert to response format
	var response []map[string]interface{}
	for _, img := range images {
		response = append(response, img.ToMap())
	}

	utils.RespondSuccess(c, response, "User media retrieved successfully")
}

// GetUserMediaHandler returns all media for a specific user (public endpoint)
func GetUserMediaHandler(c *gin.Context) {
	userIDStr := c.Param("user_id")
	userID, err := strconv.ParseUint(userIDStr, 10, 32)
	if err != nil {
		utils.RespondError(c, "Invalid user ID", 400)
		return
	}

	var images []models.Image
	if err := conf.DB.Where("user_id = ? AND is_active = ?", uint(userID), true).
		Order("is_profile DESC, created_at DESC").Find(&images).Error; err != nil {
		log.Printf("Failed to fetch user media: %v", err)
		utils.RespondError(c, "Failed to fetch media", 500)
		return
	}

	// Convert to response format (only public fields)
	var response []map[string]interface{}
	for _, img := range images {
		data := img.ToMap()
		// Remove sensitive fields for public access
		delete(data, "file_path")
		response = append(response, data)
	}

	utils.RespondSuccess(c, response, "User media retrieved successfully")
}

// SetProfileImageHandler sets an image as the user's profile picture
func SetProfileImageHandler(c *gin.Context) {
	// Get user ID from context
	userID, exists := c.Get("user_id")
	if !exists {
		utils.RespondError(c, "Authentication required", 401)
		return
	}

	var request struct {
		ImageID uint `json:"image_id" binding:"required"`
	}

	if err := c.ShouldBindJSON(&request); err != nil {
		utils.RespondError(c, "Invalid request data", 400)
		return
	}

	// Start transaction
	tx := conf.DB.Begin()

	// First, unset all profile images for this user
	if err := tx.Model(&models.Image{}).
		Where("user_id = ?", userID).
		Update("is_profile", false).Error; err != nil {
		tx.Rollback()
		log.Printf("Failed to unset profile images: %v", err)
		utils.RespondError(c, "Failed to update profile image", 500)
		return
	}

	// Then set the selected image as profile
	var image models.Image
	if err := tx.Where("id = ? AND user_id = ? AND is_active = ?",
		request.ImageID, userID, true).First(&image).Error; err != nil {
		tx.Rollback()
		utils.RespondError(c, "Image not found", 404)
		return
	}

	image.IsProfile = true
	if err := tx.Save(&image).Error; err != nil {
		tx.Rollback()
		log.Printf("Failed to set profile image: %v", err)
		utils.RespondError(c, "Failed to update profile image", 500)
		return
	}

	// Commit transaction
	tx.Commit()

	utils.RespondSuccess(c, image.ToMap(), "Profile image updated successfully")
}